### PR TITLE
Adding Envoy bootstrap template for a custom Pilot implementation.

### PR DESCRIPTION
**Pull request by alexburnos**
New template connects to Pilot using Google gRPC Envoy client, which
allows to perform authz by passing additional credentials. Placed into
install/gcp due to being GCP installation specific.

To enable this template, introducing {{ .discovery_address }} variable,
which passes --discoveryAddress flag value "as is", without splitting it into
address/port_value parts as currently done for the {{ .pilot_grpc_address }} variable.

### CODE REVIEW

**Comment by alexburnos** [comment]
Work towards #11336

File: install/gcp/bootstrap/gcp_envoy_bootstrap.json [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        "address": {
+          "socket_address": {
+            "address": "0.0.0.0",
+            "port_value": 15001
+          }
+        },
+        "use_original_dst": true,
+        "filter_chains": [
+          {
+            "filters": [
```
Comment (rshriram) [inline install/gcp/bootstrap/gcp_envoy_bootstrap.json]: why http ? its safer to have TCP proxy here.

File: install/gcp/bootstrap/gcp_envoy_bootstrap.json [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        "name": "virtual",
+        "address": {
+          "socket_address": {
+            "address": "0.0.0.0",
+            "port_value": 15001
+          }
+        },
+        "use_original_dst": true,
+        "filter_chains": [
+          {
```
Comment (costinm) [inline install/gcp/bootstrap/gcp_envoy_bootstrap.json]: Would be good to have some comments on what's different from the normal template.

I think we discussed in the past - having a separate file is safer, but keeping in sync may be trickier. If it's only the google grpc stuff - maybe we can add it in the main template as an option ?

Comment (alexburnos) [reply install/gcp/bootstrap/gcp_envoy_bootstrap.json]: You are right, maybe just adding an option would be better. Let me try it and either come back with justification or another PR.

File: install/gcp/bootstrap/gcp_envoy_bootstrap.json (hunk 1 above)
Comment (alexburnos) [reply install/gcp/bootstrap/gcp_envoy_bootstrap.json]: I believe this is similar to default Istio template? Nevertheless, hold for a bit. I might send different PR (see comment from @costinm above).

File: install/gcp/bootstrap/gcp_envoy_bootstrap.json (hunk 2 above)
Comment (costinm) [reply install/gcp/bootstrap/gcp_envoy_bootstrap.json]: I would rather merge this PR in 1.1, and have another PR in master. At least this one is very safe.

**Comment by alexburnos** [comment]
> /lgtm
> /approve
> 
> Change is safe - long term we'll probably want to merge the templates ( since google grpc can be used in other envionments), but for 1.1 this seems the safest option.
> 
> Maybe we should have an option for 'call_credentials' - and replace 'gcp' with 'google_grpc' - that's the only thing specific to gcp. But 1.2

Thank you. Sorry, didn't see this reply while working on https://github.com/istio/istio/pull/11431 which re-uses existing template.

For 1.1, which one would you prefer to merge? This one or #11431 that re-uses the template?

**Comment by costinm** [comment]
You need to resolve conflicts - let's merge this as it's safer. Note there is another related PR https://github.com/istio/istio/issues/11458  where the config will be in a configmap, fully customizable by user.

**Comment by duderino** [comment]
@alexburnos please fix the failing tests asap so we can get this into 1.1?

**Comment by duderino** [comment]
From https://circleci.com/gh/istio/istio/323222?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

```
?   	istio.io/istio/security/tools/generate_cert	[no test files]
# istio.io/istio/tests/integration2/pilot [istio.io/istio/tests/integration2/pilot.test]
tests/integration2/pilot/sidecar_api_test.go:45:3: unknown field 'DNSDomain' in struct literal of type "istio.io/istio/pilot/pkg/model".Proxy
FAIL	istio.io/istio/tests/integration2/pilot [build failed]
# istio.io/istio/tests/integration2/pilot [istio.io/istio/tests/integration2/pilot.test]
tests/integration2/pilot/sidecar_api_test.go:45:3: unknown field 'DNSDomain' in struct literal of type "istio.io/istio/pilot/pkg/model".Proxy
FAIL	istio.io/istio/tests/integration2/pilot [build failed]
# istio.io/istio/tests/integration2/pilot [istio.io/istio/tests/integration2/pilot.test]
tests/integration2/pilot/sidecar_api_test.go:45:3: unknown field 'DNSDomain' in struct literal of type "istio.io/istio/pilot/pkg/model".Proxy
FAIL	istio.io/istio/tests/integration2/pilot [build failed]
ok  	istio.io/istio/tests/integration2/citadel	0.102s	coverage: 1.5% of statements in istio.io/istio/...
# istio.io/istio/tests/integration2/pilot [istio.io/istio/tests/integration2/pilot.test]
tests/integration2/pilot/sidecar_api_test.go:45:3: unknown field 'DNSDomain' in struct literal of type "istio.io/istio/pilot/pkg/model".Proxy
FAIL	istio.io/istio/tests/integration2/pilot [build failed]
istio.io/istio/tests/integration2/pilot
```

**Comment by alexburnos** [comment]
I see there were some recent changes in release-1.1 related to pilot/proxy since my last merge yesterday. Merged again just now. Let's try to re-test again. Thank you.

**Pull request merged**